### DISCUSSION TITLE
To integrate Modified LZ4 and ZSTD code in ClickHouse

### DISCUSSION THREAD

**Discussion by raj-rahul51** [ClickHouse/ClickHouse#58737] [Q&A] [answered]
How would I integrate my LZ4 and ZSTD code into ClickHouse build system?
Is there any path need to be set in CMAKE build? or Is there any flag that I can pass while compiling ClickHouse?

Thanks in Advance

**Comment by alexey-milovidov** [comment, marked as answer]
There are two possibilities:

1. It is published as a library, similarly to lz4 and zstd.

Then you should add it by analogy with other libraries:
- add a submodule or a folder with the source code to `contrib` directory;
- add a nearby library-name-cmake directory and write a CMakeLists, describing how to build its translation units into a static library;
- add this library as a dependency in other CMakeLists, similarly to how lz4 and zstd is added;

2. It is just a pair of .h/.cpp/.c files.

Then, add them in a subdirectory in the `base` directory, add it to the base's CMakeLists, and use it where appropriate.

**Reply by raj-rahul51** [reply]
Thank you Alexey!

I have followed your first steps to integrate ZSTD, with my code, but in one file src/Compression/CompressionInfo.h in below function what value should I set for Mod_ZSTD which is highlighted in Bold.

enum class CompressionMethodByte : uint8_t
{
    NONE            = 0x02,
    LZ4             = 0x82,
    ZSTD            = 0x90,
    **Mod_ZSTD    = ,**
    Multiple        = 0x91,
    Delta           = 0x92,
    T64             = 0x93,
    DoubleDelta     = 0x94,
    Gorilla         = 0x95,
    AES_128_GCM_SIV = 0x96,
    AES_256_GCM_SIV = 0x97,
    FPC             = 0x98,
    DeflateQpl      = 0x99,
    GCD             = 0x9a,
};

2. Do I need to add new file in path src/Compression/ something like **CompressionCodecMod_ZSTD.cpp** to pass my code handles to match with the one present by default **CompressionCodecZSTD.cpp**

**Reply by alexey-milovidov** [reply]
If your version of ZSTD uses the same (compatible) format, it should use the same value as ZSTD, and we can simply use it in place of zstd if it is always better.

If your version of ZSTD uses an incompatible format, choose the next available value, e.g. 0x9b.

**Reply by alexey-milovidov** [reply]
If your library is a drop-in replacement for ZSTD, just use it in place of ZSTD.
If not, you will have to make your own implementation derived from ICompressionCodec.

**Reply by raj-rahul51** [reply]
Thank you Alexey!

I have successfully integrated LZ4 and ZSTD facing some challenges, that I will try to solve it.

I have question on Benchmarking of LZ4\ZSTD - 
1 - Is clickhouse-benchmark need to be used for benchmarking or any other tool?
2 - Which data I can capture and Is lesser is better or greater is better for below parameters?
QPS
RPS
result RPS
result MiB/s

for example by random numbers

ALGO                         LZ4                       Mod_LZ4
QPS                      ->1.7	                ->1.6
RPS                      ->106904962.4	->100167698.5
result RPS            -> 89087468.63	->83409882.41
result MiB/s          ->679.6	        -> 636.3

Thanks in Advance

**Reply by UnamedRus** [reply]
All of them: greater is better.

QPS is queries per second
RPS is rows per second

MiB/s is MiB processed per second

**Reply by raj-rahul51** [reply]
Thank you @UnamedRus
